<h2 style='color:blue' align='center'>L1 and L2 Regularization</h2>

In [1]:
# import libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns

In [2]:
# Suppress Warnings for clean notebook
import warnings
warnings.filterwarnings('ignore')

**We are going to use Melbourne House Price Dataset where we'll predict House Predictions based on various features.**
#### The Dataset Link is
https://www.kaggle.com/anthonypino/melbourne-housing-market

In [3]:
# read dataset
dataset = pd.read_csv('https://raw.githubusercontent.com/codebasics/py/master/ML/16_regularization/Melbourne_housing_FULL.csv')

In [4]:
dataset.head()

Suburb             Address  ...             Regionname Propertycount
0  Abbotsford       68 Studley St  ...  Northern Metropolitan        4019.0
1  Abbotsford        85 Turner St  ...  Northern Metropolitan        4019.0
2  Abbotsford     25 Bloomburg St  ...  Northern Metropolitan        4019.0
3  Abbotsford  18/659 Victoria St  ...  Northern Metropolitan        4019.0
4  Abbotsford        5 Charles St  ...  Northern Metropolitan        4019.0

[5 rows x 21 columns]

In [7]:
dataset.nunique()
dataset.shape

(34857, 21)

In [9]:
# let's use limited columns which makes more sense for serving our purpose - filter methods can be used to filter data meaningfully
cols_to_use = ['Suburb', 'Rooms', 'Type', 'Method', 'SellerG', 'Regionname', 'Propertycount', 
               'Distance', 'CouncilArea', 'Bedroom2', 'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'Price']
dataset = dataset[cols_to_use]

In [10]:
dataset.head()

Suburb  Rooms Type Method  ...  Car Landsize  BuildingArea      Price
0  Abbotsford      2    h     SS  ...  1.0    126.0           NaN        NaN
1  Abbotsford      2    h      S  ...  1.0    202.0           NaN  1480000.0
2  Abbotsford      2    h      S  ...  0.0    156.0          79.0  1035000.0
3  Abbotsford      3    u     VB  ...  1.0      0.0           NaN        NaN
4  Abbotsford      3    h     SP  ...  0.0    134.0         150.0  1465000.0

[5 rows x 15 columns]

In [11]:
dataset.shape

(34857, 15)

#### Checking for Nan values

In [12]:
dataset.isna().sum()

Suburb               0
Rooms                0
Type                 0
Method               0
SellerG              0
Regionname           3
Propertycount        3
Distance             1
CouncilArea          3
Bedroom2          8217
Bathroom          8226
Car               8728
Landsize         11810
BuildingArea     21115
Price             7610
dtype: int64

#### Handling Missing values

In [13]:
# Some feature's missing values can be treated as zero (another class for NA values or absence of that feature)
# like 0 for Propertycount, Bedroom2 will refer to other class of NA values
# like 0 for Car feature will mean that there's no car parking feature with house
cols_to_fill_zero = ['Propertycount', 'Distance', 'Bedroom2', 'Bathroom', 'Car']
dataset[cols_to_fill_zero] = dataset[cols_to_fill_zero].fillna(0)

# other continuous features can be imputed with mean for faster results since our focus is on Reducing overfitting
# using Lasso and Ridge Regression
dataset['Landsize'] = dataset['Landsize'].fillna(dataset.Landsize.mean())
dataset['BuildingArea'] = dataset['BuildingArea'].fillna(dataset.BuildingArea.mean())

**Drop NA values of Price, since it's our predictive variable we won't impute it**

In [14]:
dataset.dropna(inplace=True)

In [15]:
dataset.shape

(27244, 15)

#### Let's one hot encode the categorical features, get dummy variable and avoid dummy variable trap

In [16]:
dataset = pd.get_dummies(dataset, drop_first=True)

In [17]:
dataset.head()

Rooms  ...  CouncilArea_Yarra Ranges Shire Council
1      2  ...                                       0
2      2  ...                                       0
4      3  ...                                       0
5      3  ...                                       0
6      4  ...                                       0

[5 rows x 745 columns]

#### Let's bifurcate our dataset into train and test dataset

In [18]:
X = dataset.drop('Price', axis=1)
y = dataset['Price']

In [19]:
from sklearn.model_selection import train_test_split
train_X, test_X, train_y, test_y = train_test_split(X, y, test_size=0.3, random_state=2)

#### Let's train our Linear Regression Model on training dataset and check the accuracy on test set

In [20]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(train_X, train_y)

In [21]:
reg.score(test_X, test_y)

0.13853683161569574

In [22]:
reg.score(train_X, train_y)

0.6827792395792723

**Here training score is 68% but test score is 13.85% which is very low**

<h4 style='color:purple'>Normal Regression is clearly overfitting the data, let's try other models</h4>

#### Using Lasso (L1 Regularized) Regression Model

In [39]:
from sklearn import linear_model
lasso_reg = linear_model.Lasso(alpha=60, max_iter=100, tol=0.1)
lasso_reg.fit(train_X, train_y)

Lasso(alpha=60, max_iter=100, tol=0.1)

In [40]:
lasso_reg.score(test_X, test_y)

0.673098533487691

In [41]:
lasso_reg.score(train_X, train_y)

0.6753779363632062

#### Using Ridge (L2 Regularized) Regression Model

In [46]:
from sklearn.linear_model import Ridge
ridge_reg= Ridge(alpha=70, max_iter=100, tol=0.1)
ridge_reg.fit(train_X, train_y)

Ridge(alpha=70, max_iter=100, tol=0.1)

In [47]:
ridge_reg.score(test_X, test_y)

0.6634879830338729

In [48]:
ridge_reg.score(train_X, train_y)

0.6576352251691995

**We see that Lasso and Ridge Regularizations prove to be beneficial when our Simple Linear Regression Model overfits. These results may not be that contrast but significant in most cases.Also that L1 & L2 Regularizations are used in Neural Networks too**